### ISSUE TITLE
Initialization of recurrent weight for the adding problem or similar problems

### ISSUE THREAD

**Issue by Sunnydreamrain** [batzner/indrnn#9]
Hi,

Thanks for the implementation. 
I have one comment on "why IndRNN on 5000 time steps does not work in your implementation". I think it is related to the initialization of the recurrent weight. 
As shown in the paper, to keep long-term memory, the recurrent weight needs to be around 1. For the adding problem, and for the last IndRNN layer, only the last output is useful. Therefore, there is no need to keep the short-term memory.  Accordingly, the recurrent weight of the last IndRNN layer can be initialized to be all 1 or a range (1-epsilon, 1+epsilon) where epsilon is small. By the way, for relu, the recurrent weight for the other layers can be initialized to (0, recurr_max) without the negative part, and it is initialized with Uniform distribution to keep all kinds of memory.

In your implementation, only 128 units are used. With the uniform distribution for the last IndRNN layer, the number of units that can keep long-term memory is very small, thus making it very hard to solve tasks of long sequences.

This applies to other tasks that only require the final output such as mnist classification and action recognition.

Could you please give it a try? It works on my end.

Thanks.

**Comment by batzner** [comment]
Thank you for the elaborate explanation - it makes sense to me. I'm trying it out right now. Some questions:

1. With the recurrent weight being around initially 1 for all cells, the initial MSE explodes in the first step. This is probably due to every unit keeping its hidden state and adding the input to it at each time step. How did you overcome this problem?
2. Why is `(0, recurr_max)` sufficient for ReLU? In the paper, they state that they used `(-recurr_max, recurr_max)`

**Comment by Sunnydreamrain** [comment]
> With the recurrent weight being around initially 1 for all cells, the initial MSE explodes in the first step. This is probably due to every unit keeping its hidden state and adding the input to it at each time step. How did you overcome this problem?

You mean "initialize the recurrent weights to 1 for all IndRNN layers or just the last IndRNN layer?" Just to be clear, I meant only the last IndRNN layer.
If it is only for the last IndRNN layer, it wouldn't explode. Maybe check your input weight initialization? In my code, I used a relatively small one `Normal(0.001)` similarly as IRNN. 

> Why is (0, recurr_max) sufficient for ReLU? In the paper, they state that they used (-recurr_max, recurr_max)

In the paper, it only says "constraining it to `(-recurr_max, recurr_max)`". I think the initialization to a proper range as `(0, recurr_max)` is fine.

My code is implemented in Theano and Lasagne. I was on a trip in the last few days and just got back. I will upload my codes to Github as well. We can cross-check them.

Thanks.

**Comment by Sunnydreamrain** [comment]
Hi,

I have uploaded my code to Github. It is based on Theano and Lasagne. I will keep updating it.  
https://github.com/Sunnydreamrain/IndRNN_Theano_Lasagne

I have also listed some considerations of the implementation in the end. Hope it can be useful to yours.  
Let me know if you find any problems.  

Thanks.

**Comment by batzner** [comment]
Thank you for the code. I've been running experiments since yesterday - here are my results:

Setting all recurrent weights to 1 and the input weights to `random_gaussian(mean=0.0, stddev=0.001)` (commit 7e3110199f04093ce19bcb8940d49cbe7d3aced5):
![tall](https://user-images.githubusercontent.com/10948237/37699028-cf8d1262-2ce5-11e8-98a8-28b8399509a1.png)

Same as above except for setting the recurrent weights of the first layer to `[-max_abs, max_abs]` (commit c4067f0af1cfc8dc03a1735b9789a84764a4f037):

![tall](https://user-images.githubusercontent.com/10948237/37699085-0e14aaa4-2ce6-11e8-8102-b895434a307e.png)

The special recurrent weights initializer clearly makes the net converge super-fast. Comparing these results to Figure 2 in the paper I'm not sure, which version they used. Probably the special `[0, max_abs]` or `[-max_abs, max_abs]` init for the recurrent weights in the first layer. 

It seems to me that the special init for the first layer fits the architecture specifically to the addition problem a bit. Do you think it makes sense to use this init for the first layer on other problems (Sequential MNIST, Language Modeling etc.) as well?

**Comment by Sunnydreamrain** [comment]
Hi,

The results seem nice. If you keep training and dropping the learning rate, it will converge to a much smaller MSE.  
By `first layer`, you mean the layer before the final output or the layer after the input?  
In the implementation of the paper (which is also shown in my Git), I used [0, max_abs]. I did not test a lot of initialization techniques since this one already shows a nice result. It is good to know other initializations work better. In my opinion, it should also work for other tasks as well. But this is only one test yet and one more example showing this would be better.

Thanks.

**Comment by batzner** [comment]
The `T=5000` have finally finished running (took 2 days). I updated the figures above. The learning curves for a uniform `[-max_abs, max_abs]` init for the first layer (after the input) look really similar to the ones in the paper. So, thank you very much for your input - it was very helpful for "solving" the addition problem. Now, I will try the Sequential MNIST. Did you already run experiments on that?

**Comment by Sunnydreamrain** [comment]
Thanks for the update.  

> Now, I will try the Sequential MNIST. Did you already run experiments on that?
Hi, I am the author of the paper and the results presented in the paper are obtained from the code I uploaded. 
[https://github.com/Sunnydreamrain/IndRNN_Theano_Lasagne](https://github.com/Sunnydreamrain/IndRNN_Theano_Lasagne)

I have also updated the example of the language modelling task using PTB_c.  

Let me know if you have any issues in getting the right results.

Again, thanks for your implementation.

**Comment by batzner** [comment]
> Hi, I am the author of the paper

Oh, sorry I didn't catch that 😄. I referenced your repo in the Readme so that people know where to find the original code. Thank you very much for uploading the implementation! It will make it a lot easier to reproduce the experiments in TensorFlow.

**Comment by Sunnydreamrain** [comment]
Great. Thanks a lot for the reference.

**Comment by Sunnydreamrain** [comment]
Hi,

In our discussion, you mentioned `a uniform [-max_abs, max_abs] init for the first layer (after the input) ` seems to converge faster. Well, I have tried in several experiments (mnist and PTB) and it was no better than `uniform [0, max_abs]`.

Let me know if you find it otherwise in other experiments.

Thanks.

**Comment by batzner** [comment]
That's interesting. I only tried the `uniform [-max_abs, max_abs]` vs. setting all recurrent weights to 1 in the first layer. The uniform performed much better but resulted in a similar convergence to the figures in the paper. So, it seems to make no difference whether you use `uniform [-max_abs, max_abs]` or `uniform [0, max_abs]` as long as you don't set them all to 1 in the first layer.

**Comment by Sunnydreamrain** [comment]
> So, it seems to make no difference whether you use uniform [-max_abs, max_abs] or uniform [0, max_abs] as long as you don't set them all to 1 in the first layer.  

Well, that is hard to say. For the adding problem, it is quite simple. As shown in the paper, only one neuron in the second layer could solve it and other neurons could be dead all the time. However, for other difficult problems, more neurons are needed. So it would be better to initialize the neurons in a way that keeps more neurons activated and efficient.

**Comment by batzner** [comment]
Yes, I only meant the addition problem specifically. For real world tasks, the initial "memory" of the neurons is probably a hyper-parameter worth optimizing.

**Comment by Sunnydreamrain** [comment]
Okay.
